src/main/java/build/buildfarm/cas/cfc/CasFallbackHandler.java

```diff
@@ -0,0 +1,105 @@
+   * @param delegate The delgate to start.
+   * @param onStartPut A callback when CAS objects are put (relevant for CasFileCache).
+   * @param removeDirectoryService Service for deleting files (relevant for CasFileCache).
+   * @param skipLoad Whether to load the existing cache (relevant for CasFileCache).
+   */
+  public static void start(
+      ContentAddressableStorage delegate,
+      Consumer<Digest> onStartPut,
+      ExecutorService removeDirectoryService,
+      boolean skipLoad)
+      throws IOException, InterruptedException {
+    // start delegate if we specifically have a CASFileCache
+    if (delegate != null && delegate instanceof CASFileCache) {
+      CASFileCache fileCacheDelegate = (CASFileCache) delegate;
+      fileCacheDelegate.start(onStartPut, removeDirectoryService, skipLoad);
+    }
+  }
+  /**
+   * @brief Get an inputstream to read the given digest data.
+   * @details Creates the delegate's inputstream if possible.
+   * @param delegate The delgate to get the inputstream from.
+   * @param e Exception from primary (refactor to not use exceptions for control flow).
+   * @param digest The digest to read.
+   * @param offset The reading offset for the data to read.
+   * @return Inputstream to blob.
+   * @note Suggested return identifier: istream.
+   */
+  public static InputStream newTransparentInput(
+      ContentAddressableStorage delegate, NoSuchFileException e, Digest digest, long offset)
+      throws IOException {
+    if (delegate == null) {
+      throw e;
+    }
+    return delegate.newInput(digest, offset);
+  }
+  /**
+   * @brief Query delegate CAS to find missing blobs.
+   * @details Will not query delegate if there are no missing blobs given.
+   * @param delegate The delegate to check for missing blobs.
+   * @param missingDigests Missing blobs to check for.
+   * @return Found blobs.
+   * @note Suggested return identifier: foundBlobs.
+   */
+  public static Iterable<Digest> findMissingBlobs(
+      ContentAddressableStorage delegate, ImmutableList<Digest> missingDigests)
+      throws InterruptedException {
+    // skip calling the fallback CAS if it does not exist or we already found the digests
+    if (delegate == null || missingDigests.isEmpty()) {
+      return missingDigests;
+    }
+
+    return delegate.findMissingBlobs(missingDigests);
+  }
+  /**
+   * @brief Check if delegate CAS contains blob.
+   * @details Will be false if CAS delegate is unavailable.
+   * @param delegate The delegate to check for the digest.
+   * @param digest The digest to check for.
+   * @param result The digest results to be populated.
+   * @return Whether the digest was found.
+   * @note Suggested return identifier: found.
+   */
+  public static boolean contains(
+      ContentAddressableStorage delegate, Digest digest, Digest.Builder result) {
+    if (delegate == null) {
+      return false;
+    }
+    return delegate.contains(digest, result);
```

Comment:
Still prefer `return delegate != null && delegate.contains(digest, result);` to this.

Reply:
fixed